### PR TITLE
Bumper Script for Vbot

### LINKED ISSUE #129: Write a CLI tool to bump versions
## Requirements 
- [x] Should take a single argument `'major'|'minor'|'patch'` or a specific version.
  - [ ] Should default to `'patch'`.
  - [ ] Should verify that the new version is a valid version
    - [ ] Must be greater than the old version
    - [ ] Must not have been used before
- [x] Should update the value in `/package.json`
- [ ] #139
- [x] Should commit changes with message `'🚀🔖 release vX.X.X'`
- [x] Should create a new git tag (if on `master`?)
- [x] Should be written in TS or Ruby

## Related tools
- https://docs.npmjs.com/cli/v8/commands/npm-version

### PR DESCRIPTION
**Pull request by EdgarSaldivar**
I believe this was is specced to the issue. Im ready to be torn apart.

### CODE REVIEW

File: scripts/bump.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
+import * as fs from 'fs';
+import simpleGit, { SimpleGitOptions } from 'simple-git';
+
+// Increments the specified version based on the release type
```
Comment (dudeofawesome) [inline scripts/bump.ts]: Prefer to use /** style comments for JSDoc compat.
Also applies to other definition comments.

File: scripts/bump.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  } else {
+    parts[2]++;
+  }
+
+  return parts.join('.');
+}
+
+// Reads the current version from the package.json file
+function readPackageJson(): any {
+  const packageJson = fs.readFileSync('package.json', { encoding: 'utf-8' });
```
Comment (dudeofawesome) [inline scripts/bump.ts]: An alternative to reading in the package file would be to `import` the JSON at the top.

File: scripts/bump.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  return JSON.parse(packageJson);
+}
+
+// Writes the new version to the package.json file
+function writePackageJson(packageJson: any): void {
+  fs.writeFileSync('package.json', JSON.stringify(packageJson, null, 2), {
```
Comment (dudeofawesome) [inline scripts/bump.ts]: I'm worried this will clobber any formatting in the JSON file. Have you tested for that?

File: scripts/bump.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    encoding: 'utf-8',
+  });
+}
+
+// Bumps the version in the package.json file
+export default async function bump(): Promise<void> {
```
Comment (dudeofawesome) [inline scripts/bump.ts]: Prefer to call the main method `main`, especially for CLI apps. You might also prefer to split your CLI param parsing and handling logic into 2 methods (`main` and `bump`?)

File: package.json [hunk 5]
```diff
@@ -79,6 +83,7 @@
     "pluralize": "^8.0.0",
     "reflect-metadata": "^0.1.13",
     "rxjs": "^7.2.0",
+    "simple-git": "^3.16.0",
```
Comment (dudeofawesome) [inline package.json]: This should be a dev-dep

File: scripts/bump.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  const options: Partial<SimpleGitOptions> = {
```
Comment (dudeofawesome) [inline scripts/bump.ts]: This is only ever used once, so probably no need to define is as a local var.

File: scripts/bump.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  packageJson.version = newVersion;
+  writePackageJson(packageJson);
+  const url = `git@github.com:Virginity-Bot/virginity-bot.git`;
+
+  console.log(`Bumped version from ${currentVersion} to ${newVersion}`);
+
+  git
+    .add(`./`)
+    .commit(`🚀🔖 release v${newVersion}`)
+    .addRemote(`${branchName}`, `${url}`);
```
Comment (dudeofawesome) [inline scripts/bump.ts]: Are you sure we need this `.addRemote`? Seems like the user should already have their remote set up?

File: package.json (hunk 5 above)
Comment (EdgarSaldivar) [reply package.json]: dev-dep?

**Comment by dudeofawesome** [comment]
How would you feel about breaking this out into a separate repo / package? Would be nice to be able to reuse this for other projects easily.

**Comment by EdgarSaldivar** [comment]
Yeah once everything is patched up sounds like a good idea.

File: package.json (hunk 5 above)
Comment (dudeofawesome) [reply package.json]: dev dependency

`npm i -D simple-git`

File: package.json [hunk 8]
```diff
@@ -111,4 +116,4 @@
     "tsconfig-paths": "4.1.1",
     "typescript": "^4.7.4"
   }
-}
+}
```
Comment (dudeofawesome) [inline package.json]: Don't want to remove that trailing newline.

Comment (dudeofawesome) [reply package.json]: I think we're gonna need to rethink the method we use to update the `package.json`.

Comment (dudeofawesome) [reply package.json]: Did you look any further into that npm built-in command [1]? Might be useful here.

[1] https://docs.npmjs.com/cli/v8/commands/npm-version

Comment (EdgarSaldivar) [reply package.json]: My original thought was maybe a script that runs that command thats built in but i was under the impression that wasnt what you wanted.

File: scripts/bump.ts [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+export default async function bump(): Promise<void> {
+  const options: Partial<SimpleGitOptions> = {
+    baseDir: process.cwd(),
+    binary: 'git',
+    maxConcurrentProcesses: 6,
+    trimmed: false,
+  };
+
+  const git = simpleGit(options);
+  const release: string = process.argv.slice(2) as unknown as string;
```
Comment (dudeofawesome) [inline scripts/bump.ts]: I think we're missing the default to `patch` requirement?

File: scripts/bump.ts [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+  const packageJson = readPackageJson();
+  const currentVersion = packageJson.version;
+  const newVersion = incrementVersion(currentVersion, release);
+  const branchSummary = await git.branch();
+  const branchName = branchSummary.current;
+  packageJson.version = newVersion;
```
Comment (dudeofawesome) [inline scripts/bump.ts]: We should add a check here to verify there isn't a pre-existing tag using this version number.

File: scripts/bump.ts [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+  writePackageJson(packageJson);
+  const url = `git@github.com:Virginity-Bot/virginity-bot.git`;
+
+  console.log(`Bumped version from ${currentVersion} to ${newVersion}`);
+
```
Comment (dudeofawesome) [inline scripts/bump.ts]: Do we check that no other changes are going to be accidentally committed?

File: scripts/bump.ts (hunk 9 above)
Comment (dudeofawesome) [reply scripts/bump.ts]: Does this support a specifying a custom version number?

Comment (EdgarSaldivar) [reply scripts/bump.ts]: It does default to patch.

File: scripts/bump.ts (hunk 11 above)
Comment (EdgarSaldivar) [reply scripts/bump.ts]: We do not, I need claification on that. Like the bump should be the "only" thing being commited and therefore not execute if anything else is staged?

File: scripts/bump.ts (hunk 2 above)
Comment (EdgarSaldivar) [reply scripts/bump.ts]: huh okay maybe that would be cleaner

File: package.json (hunk 5 above)
Comment (dudeofawesome) [reply package.json:83]: Do we still depend on this?

File: package.json [hunk 12]
```diff
@@ -79,6 +80,7 @@
     "pluralize": "^8.0.0",
     "reflect-metadata": "^0.1.13",
     "rxjs": "^7.2.0",
+    "simple-git": "^3.16.0",
```
Comment (dudeofawesome) [inline package.json]: Do we still depend on this?

File: package.json [hunk 13]
```diff
@@ -103,6 +105,7 @@
     "eslint": "^8.0.1",
     "eslint-import-resolver-typescript": "^3.5.3",
     "jest": "29.3.1",
+    "libnpmversion": "^4.0.2",
```
Comment (dudeofawesome) [inline package.json]: Do we still depend on this?

File: scripts/bump.ts [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+    message: `🚀🔖 release v%s`, // message for tag and commit, replace %s with the version
+    silent: false, // passed to @npmcli/run-script to control whether it logs
+  })
+    .then((newVersion) => {
+      console.error(`version updated! ${npmVersion.newVersion}`, newVersion);
+    })
+    .catch(console.log());
+}
+
+bump();
```
Comment (dudeofawesome) [inline scripts/bump.ts]: Do we still need need this?

File: scripts/bump.sh [hunk 15]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
+
+echo "Bumping... "
+
+if [[ -z "$1" ]] ; then
+      npm version --message "$message" patch
```
Comment (dudeofawesome) [inline scripts/bump.sh]: Where is `$message` defined?

File: scripts/bump.sh [hunk 16]
```diff
@@ -0,0 +1,11 @@
...
+      exit 1
```
Comment (dudeofawesome) [inline scripts/bump.sh]: Are you sure you want to error out here?

File: scripts/bump.sh [hunk 17]
```diff
@@ -0,0 +1,11 @@
...
+else
+  npm version $1 -m "$message"
+  exit 1
```
Comment (dudeofawesome) [inline scripts/bump.sh]: ditto

File: scripts/bump.sh [hunk 18]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
+
+echo "Bumping... "
+
+if [[ -z "$1" ]] ; then
+      npm version --message "$message" patch
+      exit 1
+else
+  npm version $1 -m "$message"
```
Comment (dudeofawesome) [inline scripts/bump.sh]: I think the message needs to come before the bump param?

**Review by dudeofawesome — CHANGES_REQUESTED** [review]

File: scripts/bump.sh [hunk 19]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+
+echo "Bumping... "
+message='🚀🔖 release v%s'
+
+if [[ -z "$1" ]] ; then
+      npm version --message "$message" patch
+      exit 0
```
Comment (dudeofawesome) [inline scripts/bump.sh]: check indentation

Comment (dudeofawesome) [reply scripts/bump.sh]: also, this could probably be refactored to have only a single call to `npm version` by just settings `$1` to `patch` here.

**Review by dudeofawesome — APPROVED** [review]

**Pull request merged**